src/SignalR/server/Core/src/Internal/Utf8HashLookup.cs

```diff
@@ -81,32 +109,52 @@ private bool TryGetValue(ReadOnlySpan<char> key, [MaybeNullWhen(false), AllowNul
         return false;
     }
 
-    private static int GetKeyHashCode(ReadOnlySpan<char> key)
+    private static int GetHashCode(ReadOnlySpan<char> value) =>
+        HashCodeMask & string.GetHashCode(value, StringComparison.OrdinalIgnoreCase);
+
+    private static int GetCaseSensitiveHashCode(ReadOnlySpan<byte> encodedValue)
     {
-        return HashCodeMask & string.GetHashCode(key, StringComparison.OrdinalIgnoreCase);
+        var hashCode = new HashCode();
+        hashCode.AddBytes(encodedValue);
+        return HashCodeMask & hashCode.ToHashCode();
     }
 
     private void Resize()
     {
-        var newSize = checked(count * 2 + 1);
-        var newBuckets = new int[newSize];
+        var newSize = checked(_count * 2 + 1);
         var newSlots = new Slot[newSize];
-        Array.Copy(slots, newSlots, count);
-        for (int i = 0; i < count; i++)
+
+        var newBuckets = new int[newSize];
+        var newCaseSensitiveBuckets = new int[newSize];
+
+        Array.Copy(_slots, newSlots, _count);
+
+        for (int i = 0; i < _count; i++)
         {
             int bucket = newSlots[i].hashCode % newSize;
             newSlots[i].next = newBuckets[bucket] - 1;
             newBuckets[bucket] = i + 1;
+
+            int caseSensitiveBucket = newSlots[i].caseSensitiveHashCode % newSize;
+            newSlots[i].caseSensitiveNext = newCaseSensitiveBuckets[caseSensitiveBucket] - 1;
+            newCaseSensitiveBuckets[caseSensitiveBucket] = i + 1;
         }
-        buckets = newBuckets;
-        slots = newSlots;
+
+        _slots = newSlots;
+
+        _buckets = newBuckets;
+        _caseSensitiveBuckets = newCaseSensitiveBuckets;
     }
 
-    internal struct Slot
+    private struct Slot
     {
         internal int hashCode;
-        internal int next;
-        internal string key;
+        internal int caseSensitiveHashCode;
+
         internal string value;
+        internal Memory<byte> encodedValue;
```

Comment:
why not make this `byte[]`?

Reply:
I figured `Memory<byte>.Span` was probably faster than `AsSpan<T>(this T[] array)`, but thinking about it more, I was probably mistaken considering `Memory<byte>` also has to deal with slicing and possible `MemoryManager<byte>` objects. I doubt it's a huge difference, but we might as well remove the to extra `int`s from `Slot`.

I'll change it to `byte[]`.